pkg/engine/shield/absorb.go

```diff
@@ -6,32 +6,51 @@ import (
 )
 
 func (mgr *Manager) AbsorbDamage(target key.TargetID, damage float64) float64 {
-	// TODO: ACTUAL SHIELD LOGIC GOES HERE
-	// 1. for every shield on this target, attempt to deal "damage" amount of damage to the shield
-	// 2. for each shield, a ShieldChangeEvent should be emitted (how much damage was done to the shield)
-	// 3. if shield HP <= damage, remove that shield (mgr.RemoveShield to generate event)
-	// 4. if all shields HP <= damage, return will be damage - max(ShieldHP) [the damage not blocked by shield]
-	// 5. if there are shields which HP > damage, return should be 0
-	// 6. remaining state should be only shields that starting HP > damage
+	// Check if incoming damage = 0
+	if damage != 0 {
+		var removedShields []*Instance
+		i := 0
+		// If not 0, then run loop to apply damage to all shields
+		for _, shield := range mgr.targets[target] {
+			shield.hp -= damage
+			if shield.hp < 0 {
+				shield.hp = 0
+			}
+			// Generate list of shields to remove
+			if shield.hp == 0 {
+				removedShields = append(removedShields, shield)
+			} else {
+				mgr.targets[target][i] = shield
+				i++
+			}
+		}
+		mgr.targets[target] = mgr.targets[target][:i]
 
-	// placeholder just for some basic event emission
-	for id, shield := range mgr.targets[target] {
-		prevHP := shield.HP
-		shield.HP -= damage
-		if shield.HP < 0 {
-			shield.HP = 0
+		// Getting outgoing damage
+		maxShield := mgr.MaxShield(target)
+		maxShieldHP := mgr.targets[target][maxShield]
+		newMaxShieldHP := maxShieldHP.hp - damage
+		damageOut := -newMaxShieldHP
+		if damageOut < 0 {
+			damageOut = 0
 		}
```

Comment:
I'm not seeing how this works. Lets say I have a 50 HP shield equipped and the `AbsorbDamage` call passes in `100` damage:
1. Loop over all shields (just this one shield) and subtract `100` damage from its HP
2. shield is removed since its HP becomes 0 (`50 - 100 < 0`)
3. Get the max shield. Because I no longer have a shield equipped, this will return no shield (so assuming `maxShield.hp = 0`)
4. Do `maxShield.hp - damage`, which returns `-100` since I currently have no shield
5. `damageOut` set to `100` (the original incoming damage) and it is ignored that the shield I had absorbed 50 damage

My suggestions are the following:
1. move this logic to inside where you subtract shield hp so you only need to loop over shields once
2. Determine your `damageOut` and `maxShieldHP` in this loop by finding the min damage out and `maxShieldHP`

For the 2nd, you can do conditions like the following:
```go
maxShieldHP := 0.0
damageOut := damage
for _, shield := range mgr.targets[target] {
  // Dim returns the maximum of x-y or 0.
  remaining := math.Dim(damage, shield.hp)
  if remaining < damageOut {
    damageOut = remaining
  }

  shield.hp = math.Dim(shield.hp, damage)
  if shield.hp > maxShieldHP {
    maxShieldHP = shield.hp
  }

  // add to remove list or insert back in targets array
}
```